src/algorithms/basic_testing.py

```diff
@@ -69,37 +69,40 @@ def algorithm_start(self) -> None:
                                    DebugLevel.BASIC)
         self.timer = Timer()
 
-    def check_terminated(self) -> None:
-        if self._services.algorithm.instance.testing != self:
-            from simulator.models.map_model import AlgorithmTerminated
-            raise AlgorithmTerminated()
+    def is_terminated(self) -> bool:
+        return self._services.algorithm.instance.testing != self
 
     def key_frame_internal(self, ignore_key_frame_skip: bool = False, only_render: List[int] = None, root_key_frame=None) -> None:
         """
         Internal key frame handler
         """
+        if self.timer is None:
+            self._services.debug.write_error("Algorithm " + str(self._services.algorithm.algorithm_type) + " hasn't started yet!")
+            return
         self.timer.pause()
 
         if root_key_frame:
             root_key_frame.instance.testing.key_frame(ignore_key_frame_skip=ignore_key_frame_skip, only_render=only_render)
-
+        
         if ignore_key_frame_skip or self.key_frame_skip_count >= self._services.settings.simulator_key_frame_skip:
             self.display_info = self._services.algorithm.instance.set_display_info()
 
             if only_render:
                 self.display_info = [self.display_info[i] for i in only_render]
 
-            if self.key_frame_condition is not None:
-                self.key_frame_condition.acquire()
-                self.requires_key_frame = True
-                self.key_frame_condition.notify()
-                self.key_frame_condition.wait()
-                self.key_frame_condition.release()
+            if self.cv is not None:
+                with self.cv:
+                    self.requires_key_frame = True
+                    self.cv.notify()
+                    self.cv.wait_for(lambda: not self.requires_key_frame or self.is_terminated())
```

Comment:
will cv ever be terminated without getting the keyframe

Reply:
Huh?

Reply:
sorry i meant self._is_terminated(), just want to make sure that all works as intended

Reply:
Yh `self.is_terminated()` can happen from what I remember and in any case it's a very cheap check